File: wallet/src/main/java/com/worldventures/wallet/util/WalletLocationsUtil.java
package com.worldventures.wallet.util;

import com.google.android.gms.maps.model.LatLng;
import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.wallet.domain.entity.lostcard.WalletCoordinates;
import com.worldventures.wallet.domain.entity.lostcard.WalletLocation;

import java.util.List;

public class WalletLocationsUtil {

   private WalletLocationsUtil() {
   }

   public static LatLng toLatLng(WalletCoordinates position) {
      return new LatLng(
            position.lat(),
            position.lng()
      );
   }

   public static WalletLocation getLatestLocation(List<WalletLocation> smartCardLocations) {
      return !smartCardLocations.isEmpty() ? applyQueryToLocationList(smartCardLocations) : null;
   }

   private static WalletLocation applyQueryToLocationList(List<WalletLocation> smartCardLocations) {
      return Queryable.from(smartCardLocations)
            .distinct()
            .sort((smartCardLocation1, smartCardLocation2) -> smartCardLocation1.createdAt()
                  .compareTo(smartCardLocation2.createdAt()))
            .last();
   }
}

File: wallet/src/main/java/com/worldventures/wallet/util/WalletLocationsUtil.kt
package com.worldventures.wallet.util

import com.google.android.gms.maps.model.LatLng
import com.worldventures.wallet.domain.entity.lostcard.WalletCoordinates
import com.worldventures.wallet.domain.entity.lostcard.WalletLocation

object WalletLocationsUtil {

   fun toLatLng(position: WalletCoordinates) = LatLng(position.lat, position.lng)

   fun getLatestLocation(smartCardLocations: List<WalletLocation>): WalletLocation? =
         if (!smartCardLocations.isEmpty()) applyQueryToLocationList(smartCardLocations) else null

   private fun applyQueryToLocationList(smartCardLocations: List<WalletLocation>): WalletLocation =
         smartCardLocations.distinct().sortedWith(Comparator { location1, location2 ->
            location1.createdAt.compareTo(location2.createdAt)
         })
               .last()
}
